// citybrain/lib/components/charts/CBCrossChart/CBCrossChart.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Chart = require('bizcharts/lib/components/Chart');

var _Chart2 = _interopRequireDefault(_Chart);

var _Interval = require('bizcharts/lib/components/TypedGeom/Interval');

var _Interval2 = _interopRequireDefault(_Interval);

var _Axis = require('bizcharts/lib/components/Axis');

var _Axis2 = _interopRequireDefault(_Axis);

var _Tooltip = require('bizcharts/lib/components/Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _Guide = require('bizcharts/lib/components/Guide');

var _Guide2 = _interopRequireDefault(_Guide);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _bind = require('classnames/bind');

var _bind2 = _interopRequireDefault(_bind);

var _CBCrossChart = require('./CBCrossChart.less');

var _CBCrossChart2 = _interopRequireDefault(_CBCrossChart);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Line = _Guide2.default.Line;

var cx = _bind2.default.bind(_CBCrossChart2.default);

var CBCrossChart = function (_Component) {
  _inherits(CBCrossChart, _Component);

  function CBCrossChart() {
    _classCallCheck(this, CBCrossChart);

    return _possibleConstructorReturn(this, (CBCrossChart.__proto__ || Object.getPrototypeOf(CBCrossChart)).apply(this, arguments));
  }

  _createClass(CBCrossChart, [{
    key: 'returnThresholdLevel',
    value: function returnThresholdLevel() {
      var _props = this.props,
          dataSource = _props.dataSource,
          thresholdLevel = _props.thresholdLevel;

      var sortDataSource = [];
      if (!dataSource || dataSource.length < 1) {} else {
        dataSource.forEach(function (item, index) {
          sortDataSource.push(item.time);
        });
        sortDataSource = sortDataSource.sort(function (a, b) {
          return b - a;
        });
        if (thresholdLevel >= sortDataSource[0]) {
          return { tickCount: 5, min: 0, max: thresholdLevel };
        } else {
          return { tickCount: 5 };
        }
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props2 = this.props,
          dataSource = _props2.dataSource,
          height = _props2.height,
          thresholdLevel = _props2.thresholdLevel,
          guideLineText = _props2.guideLineText,
          legendText = _props2.legendText;

      if (!dataSource || dataSource.length < 1) {
        return _react2.default.createElement(
          'div',
          { className: cx('empty') },
          '\u6682\u65E0\u6570\u636E'
        );
      }
      var xLabelStyle = {
        textStyle: {
          fontFamily: 'MicrosoftYaHei',
          fontSize: '10px',
          fill: '#9B9B9B'
        }
      };
      var yLabelStyle = {
        textStyle: {
          fontFamily: 'MicrosoftYaHei',
          fontSize: '10px',
          fill: '#9B9B9B'
        }
      };
      var xLineStyle = {
        stroke: 'rgba(255,255,255,0.15)',
        lineDash: [3, 3, 3],
        lineWidth: 1
      };
      var scale = {
        date: {
          type: 'timeCat',
          tickCount: 5,
          mask: 'MM/DD'
        },
        time: this.returnThresholdLevel()
      };
      return _react2.default.createElement(
        'div',
        { className: cx('cross-chart-container') },
        _react2.default.createElement(
          'span',
          { className: cx('unit') },
          this.props.unit || '小时'
        ),
        _react2.default.createElement(
          'div',
          { className: cx('legend') },
          _react2.default.createElement(
            'div',
            { className: cx('legend-item') },
            _react2.default.createElement('p', { className: cx('legend-color') }),
            _react2.default.createElement(
              'span',
              { className: cx('legend-text'), title: legendText },
              legendText
            )
          ),
          _react2.default.createElement(
            'div',
            { className: cx('legend-item') },
            _react2.default.createElement('p', { className: cx('legend-line') }),
            _react2.default.createElement(
              'span',
              { className: cx('legend-text'), title: guideLineText },
              guideLineText
            )
          )
        ),
        _react2.default.createElement(
          _Chart2.default,
          {
            height: height || 150,
            data: dataSource,
            forceFit: true,
            padding: this.props.padding || [10, 30, 30, 40],
            scale: scale,
            animate: false },
          _react2.default.createElement(_Axis2.default, {
            name: 'date',
            grid: null,
            label: xLabelStyle,
            line: xLineStyle,
            tickLine: {
              length: 0
            } }),
          _react2.default.createElement(_Axis2.default, { name: 'time', grid: null, label: yLabelStyle }),
          _react2.default.createElement(_Tooltip2.default, null),
          _react2.default.createElement(_Interval2.default, { position: 'date*time', size: 2 // 柱状图大小粗细
            , color: ['time', function (time) {
              if (time > thresholdLevel) {
                return 'rgba(236, 32, 32, 1)';
              } else {
                return 'rgba(143, 193, 56, 1)';
              }
            }], tooltip: ['date*time', function (data, time) {
              return {
                title: data,
                name: legendText,
                value: '' + time.toFixed(2) + (_this2.props.unit || '小时')
              };
            }] }),
          _react2.default.createElement(
            _Guide2.default,
            null,
            _react2.default.createElement(Line, {
              type: 'line',
              top: true,
              start: function start(xScale, yScale) {
                var y = 100 - thresholdLevel / yScale.time.ticks[yScale.time.ticks.length - 1] * 100;
                var start = ['0%', y + '%'];
                return start;
              },
              end: function end(xScale, yScale) {
                var y = 100 - thresholdLevel / yScale.time.ticks[yScale.time.ticks.length - 1] * 100;
                var end = ['100%', y + '%'];
                return end;
              },
              lineStyle: {
                stroke: '#FF8500' }
            })
          )
        )
      );
    }
  }]);

  return CBCrossChart;
}(_react.Component);

CBCrossChart.propTypes = {
  thresholdLevel: _propTypes2.default.number,
  dataSource: _propTypes2.default.array,
  height: _propTypes2.default.number
};

exports.default = CBCrossChart;
