// citybrain/lib/components/standard/CBTab/CBTab.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _CBTab = require('./CBTab.less');

var _CBTab2 = _interopRequireDefault(_CBTab);

var _bind = require('classnames/bind');

var _bind2 = _interopRequireDefault(_bind);

var _tabs = require('antd/lib/tabs');

var _tabs2 = _interopRequireDefault(_tabs);

var _radio = require('antd/lib/radio');

var _radio2 = _interopRequireDefault(_radio);

require('antd/lib/tabs/style');

require('antd/lib/radio/style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TabPane = _tabs2.default.TabPane;
var cx = _bind2.default.bind(_CBTab2.default);

var CBTab = function (_React$Component) {
  _inherits(CBTab, _React$Component);

  function CBTab(props) {
    _classCallCheck(this, CBTab);

    var _this = _possibleConstructorReturn(this, (CBTab.__proto__ || Object.getPrototypeOf(CBTab)).call(this, props));

    var _this$props = _this.props,
        defaultSelected = _this$props.defaultSelected,
        tabList = _this$props.tabList;

    if (defaultSelected === undefined) {
      defaultSelected = tabList[0].key;
    }
    _this.state = { activeTab: defaultSelected };
    return _this;
  }

  _createClass(CBTab, [{
    key: 'handler',
    value: function handler(e) {
      var onChange = this.props.onChange;

      this.setState({ activeTab: e.target.value });
      onChange && onChange(e.target.value);
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      var contentHeight = this.props.contentHeight;

      var ele = document.getElementsByClassName('ant-tabs-content')[0];
      ele && (ele.style.height = contentHeight + 'px');
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          size = _props.size,
          title = _props.title,
          _onChange = _props.onChange,
          tabList = _props.tabList,
          type = _props.type;
      var activeTab = this.state.activeTab;

      var activeContent = tabList.find(function (item) {
        return item.key === activeTab;
      });
      switch (type) {
        // 胶囊 tab
        case 'capsule':
          return _react2.default.createElement(
            'div',
            { className: cx('capsule-tab') },
            _react2.default.createElement(
              _radio2.default.Group,
              { value: activeTab, onChange: function onChange(e) {
                  return _this2.handler(e);
                } },
              tabList && tabList.length > 0 && tabList.map(function (item) {
                return _react2.default.createElement(
                  _radio2.default.Button,
                  { value: item.key, key: item.key },
                  item.text
                );
              })
            ),
            _react2.default.createElement(
              'div',
              { className: 'ant-tabs-content' },
              activeContent.content
            )
          );
        // 一级,二级 tab
        default:
          return _react2.default.createElement(
            'div',
            { className: cx({ 'cb-tab': true, large: size === 'large', title: title && title.length > 0 }) },
            _react2.default.createElement(
              'span',
              null,
              title && title.length > 12 ? title.substr(0, 12) + '...' : title
            ),
            _react2.default.createElement(
              _tabs2.default,
              { defaultActiveKey: activeTab, animated: false, onChange: function onChange(key) {
                  return _onChange && _onChange(key);
                } },
              tabList && tabList.length > 0 && tabList.map(function (item, index) {
                return _react2.default.createElement(
                  TabPane,
                  { tab: item.text, key: item.key },
                  item.content
                );
              })
            )
          );
      }
    }
  }]);

  return CBTab;
}(_react2.default.Component);

exports.default = CBTab;


CBTab.propTypes = {
  size: _propTypes2.default.string,
  title: _propTypes2.default.string,
  defaultSelected: _propTypes2.default.string,
  contentHeight: _propTypes2.default.string,
  tabList: _propTypes2.default.array.isRequired,
  onChange: _propTypes2.default.func
};
